hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileReader.java

```diff
@@ -0,0 +1,178 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.io.hfile;
+
+import org.apache.hadoop.fs.FSDataInputStream;
+
+import java.io.ByteArrayInputStream;
+import java.io.DataInputStream;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * A reader reading a HFile.
+ */
+public class HFileReader {
+  private final FSDataInputStream stream;
+  private final long fileSize;
+  private boolean isMetadataInitialized = false;
+  private HFileContext context;
+  private List<BlockIndexEntry> blockIndexEntryList;
+  private HFileBlock metaIndexBlock;
+  private HFileBlock fileInfoBlock;
+
+  public HFileReader(FSDataInputStream stream, long fileSize) {
+    this.stream = stream;
+    this.fileSize = fileSize;
+  }
+
+  /**
+   * Initializes the metadata by reading the "Load-on-open" section.
+   *
+   * @throws IOException upon error.
+   */
+  public void initializeMetadata() throws IOException {
+    assert !this.isMetadataInitialized;
+
+    // Read Trailer (serialized in Proto)
+    HFileTrailer trailer = readTrailer(stream, fileSize);
+    this.context = HFileContext.builder()
+        .compressAlgo(trailer.getCompressionCodec())
+        .build();
+    // At this point, HFileContext is not known yet until we parse the trailer
+    HFileBlockReader blockReader = new HFileBlockReader(
+        context, stream, trailer.getLoadOnOpenDataOffset(), fileSize - HFileTrailer.getTrailerSize());
+    HFileRootIndexBlock dataIndexBlock =
+        (HFileRootIndexBlock) blockReader.nextBlock(HFileBlockType.ROOT_INDEX);
+    this.blockIndexEntryList = dataIndexBlock.readDataIndex(trailer.getDataIndexCount());
+    this.metaIndexBlock = blockReader.nextBlock(HFileBlockType.ROOT_INDEX);
+    this.fileInfoBlock = blockReader.nextBlock(HFileBlockType.FILE_INFO);
+
+    this.isMetadataInitialized = true;
+  }
+
+  /**
+   * Seeks to the key to look up.
+   *
+   * @param key Key to look up.
+   * @return The {@link KeyValue} instance in the block that contains the exact same key as the
+   * lookup key; or {@link null} if the lookup key does not exist.
+   * @throws IOException upon error.
+   */
+  public KeyValue seekTo(KeyValue key) throws IOException {
+    BlockIndexEntry lookUpKey = new BlockIndexEntry(key, -1, -1);
+    int rootLevelBlockIndex = searchBlockByKey(lookUpKey);
+    if (rootLevelBlockIndex < 0) {
+      // Key smaller than the start key of the first block
```

Comment:
Other conditions to check? Like being greater than the end of the file

Reply:
Last key is checked now.